src/orchestrator-kubernetes/src/secrets.rs

```diff
@@ -0,0 +1,93 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+//! Management of user secrets via Kubernetes.
+
+use std::iter;
+use std::sync::Arc;
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use k8s_openapi::api::core::v1::Secret;
+use k8s_openapi::ByteString;
+use kube::api::{DeleteParams, Patch, PatchParams};
+use kube::Api;
+
+use mz_repr::GlobalId;
+use mz_secrets::{SecretsController, SecretsReader};
+
+use crate::{util, KubernetesOrchestrator, FIELD_MANAGER};
+
+#[async_trait]
+impl SecretsController for KubernetesOrchestrator {
+    async fn ensure(&self, id: GlobalId, contents: &[u8]) -> Result<(), anyhow::Error> {
+        let data = iter::once(("contents".into(), ByteString(contents.into())));
+        let secret = Secret {
+            data: Some(data.collect()),
+            ..Default::default()
+        };
+        self.secret_api
+            .patch(
+                &secret_name(id),
+                &PatchParams::apply(FIELD_MANAGER).force(),
+                &Patch::Apply(secret),
+            )
+            .await?;
+        Ok(())
+    }
+
+    async fn delete(&self, id: GlobalId) -> Result<(), anyhow::Error> {
+        self.secret_api
+            .delete(&secret_name(id), &DeleteParams::default())
```

Comment:
Heads up that this does not actually wait for it to be deleted by default, nor does it handle cases where the secret already doesn't exist.

You might want to do something like what we do in https://github.com/MaterializeInc/cloud/blob/b476c2d5fe4c6c586c1efa42d0533296807b05ae/mzcloud-shared/src/k8sutil.rs#L66-L103

Reply:
Good point about not handling 404. I fixed that. I think we actually _don't_ want to wait for deletion; this happens on the coordinator main thread so better to do the fastest possible thing.

The fact that secrets can leak is a good callout. That was an intentional part of the design, but we never got around to writing the garbage collector. I filed #13579 about that and left a comment in the code. That would handle the cleanup of partially written secret files, too.

Reply:
Works for me.